File: example/java/pact_validation_with_oas/one_consumer_one_provider/consumer/src/main/java/br/com/zup/pact/consumer/service/impl/ClientServiceImpl.java
package br.com.zup.pact.consumer.service.impl;

import br.com.zup.pact.consumer.dto.BalanceDTO;
import br.com.zup.pact.consumer.dto.ClientDetailsDTO;
import br.com.zup.pact.consumer.exception.ClientNotFoundException;
import br.com.zup.pact.consumer.integration.account.service.AccountIntegrationService;
import br.com.zup.pact.consumer.repository.ClientRepository;
import br.com.zup.pact.consumer.service.ClientService;

import java.util.List;
import java.util.Optional;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class ClientServiceImpl implements ClientService {

    private final ClientRepository clientRepositoryImpl;
    private final AccountIntegrationService accountIntegrationService;

    @Override
    public Optional<ClientDetailsDTO> getClientDetails(Integer clientId) {
        return this.clientRepositoryImpl.findByClientId(clientId);
    }

    @Override
    public Optional<List<ClientDetailsDTO>> getAll() {
        return this.clientRepositoryImpl.getAll();
    }

    @Override
    public Optional<BalanceDTO> getBalance(Integer clientId) {
        final Integer accountId = this.getAccountId(clientId);
        return this.accountIntegrationService.getBalance(accountId);
    }

    private Integer getAccountId(Integer clientId) {
        return this.getClientDetails(clientId)
                .orElseThrow(() -> new ClientNotFoundException("Client with id: " + clientId + " not found!"))
                .getAccountId();
    }
}

File: example/java/kotlin/consumer/src/main/kotlin/br/com/zup/pact/consumer/service/ClientServiceImpl.kt
package br.com.zup.pact.consumer.service

import br.com.zup.pact.consumer.dto.BalanceDTO
import br.com.zup.pact.consumer.dto.ClientDetailsDTO
import br.com.zup.pact.consumer.exception.ClientNotFoundException
import br.com.zup.pact.consumer.integration.account.service.AccountIntegrationService
import br.com.zup.pact.consumer.repository.ClientRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class ClientServiceImpl(
        @Autowired val clientRepository: ClientRepository,
        @Autowired val accountIntegrationService: AccountIntegrationService
) : ClientService {

    override fun getClientDetails(clientId: Int): ClientDetailsDTO? {
        return clientRepository.findByClientId(clientId)
    }

    override fun getAll(): List<ClientDetailsDTO> {
        return clientRepository.getAll()
    }

    override fun getBalance(clientId: Int): BalanceDTO? {
        val accountId = getAccountId(clientId)
        return accountIntegrationService.getBalance(accountId)
    }

    private fun getAccountId(clientId: Int): Int {
        val clientDetails = getClientDetails(clientId).takeIf { it != null }
                ?: throw  ClientNotFoundException("Client with id: $clientId")

        return clientDetails.accountId
    }
}